solana/indexer/subscription.go

```diff
@@ -0,0 +1,219 @@
+			}
+
+			if notif == nil {
+				s.logger.Warn("received nil notification, continuing to wait for artist_coins changes")
+				continue
+			}
+			if strings.HasPrefix(notif.Channel, "artist_coins_changed") {
+				var notifData artistCoinsChangedNotification
+				err := json.Unmarshal([]byte(notif.Payload), &notifData)
+				if err != nil {
+					s.logger.Error("failed to unmarshal artist_coins changed notification", zap.Error(err))
+					continue
+				}
+				s.logger.Info("artist_coins changed, re-starting subscription",
+					zap.String("oldMint", notifData.OldMint),
+					zap.String("newMint", notifData.NewMint),
+					zap.String("operation", notifData.Operation))
+				cancel()
+				<-subCtx.Done()
+				break
+			}
+		}
+	}
+}
+
+func buildSubscriptionRequest(mintAddresses []string) (*pb.SubscribeRequest, error) {
+	commitment := pb.CommitmentLevel_CONFIRMED
+	subscription := &pb.SubscribeRequest{
+		Commitment: &commitment,
+	}
+
+	// Listen for slots for making checkpoints
+	subscription.Slots = make(map[string]*pb.SubscribeRequestFilterSlots)
+	subscription.Slots["checkpoints"] = &pb.SubscribeRequestFilterSlots{}
+
+	// Listen to all the token accounts for the mints we care about
+	subscription.Accounts = make(map[string]*pb.SubscribeRequestFilterAccounts)
+	for _, mint := range mintAddresses {
+		accountFilter := pb.SubscribeRequestFilterAccounts{
+			Owner: []string{solana.TokenProgramID.String()},
+			Filters: []*pb.SubscribeRequestFilterAccountsFilter{
+				{
+					Filter: &pb.SubscribeRequestFilterAccountsFilter_TokenAccountState{
+						TokenAccountState: true,
+					},
+				},
+				{
+					Filter: &pb.SubscribeRequestFilterAccountsFilter_Memcmp{
+						Memcmp: &pb.SubscribeRequestFilterAccountsFilterMemcmp{
+							Offset: 0,
+							Data: &pb.SubscribeRequestFilterAccountsFilterMemcmp_Base58{
+								Base58: mint,
+							},
+						},
+					},
+				},
+			},
+		}
+		subscription.Accounts[mint] = &accountFilter
+	}
+
+	// Listen to all the Audius programs for transactions (currently redundant)
+	// programs := []string{
+	// 	claimable_tokens.ProgramID.String(),
+	// 	reward_manager.ProgramID.String(),
+	// 	payment_router.ProgramID.String(),
+	// }
+	// vote := false
+	// failed := false
+	// subscription.Transactions = make(map[string]*pb.SubscribeRequestFilterTransactions)
+	// transactionFilter := pb.SubscribeRequestFilterTransactions{
+	// 	Vote:           &vote,
+	// 	Failed:         &failed,
+	// 	AccountInclude: programs,
+	// }
+	// subscription.Transactions["audiusPrograms"] = &transactionFilter
+
+	return subscription, nil
+}
+
+// Handles a message from the gRPC subscription.
+func (s *SolanaIndexer) onMessage(ctx context.Context, msg *pb.SubscribeUpdate) {
+	logger := s.logger.With(zap.String("indexerSource", "grpc"))
+
+	if slotUpdate := msg.GetSlot(); slotUpdate != nil && slotUpdate.Slot > 0 {
+		err := updateCheckpoint(ctx, s.pool, s.checkpointId, slotUpdate.Slot)
```

Comment:
So we set the slot checkpoint before we process the transactions?

Reply:
The slot subscription actually happens separately from the transactions/account updates. I don't _love_ it. With really dense transaction volume we could get away with using slots of the filtered transactions but with more sparse transactions we would easily fall way behind 3k slots and end up backfilling a lot unnecessarily (maybe that's not the worst thing)

Reply:
Just to be 100% clear - if this is a "slot update", it's _only_ a slot update. It'll contain no block, no transactions, etc. Those updates get sent separately and indexed separately. Thus checkpoint doesn't _necessarily_ mean all the transactions of the slot were indexed (which sucks) but rather that the subscription was alive and kicking and received a slot update to that slot. I _think_ that in practice these should be relatively similar in meaning, and these services might even send the slot updates last (hard to know..)

Reply:
I guess that doesn't really answer your question. If we fail to process a transaction, right now we log an error and move on.

Some more robust retry logic or a "failed tx" table might be in order.